integration_tests/test_feegrant.py

```diff
@@ -0,0 +1,272 @@
+        == fee_granter_balance - fee_coins * number_of_periods
+    )
+    assert (
+        cluster.balance(fee_grantee_address)
+        == fee_grantee_balance - transaction_coins * number_of_periods
+    )
+    assert (
+        cluster.balance(receiver_address)
+        == receiver_balance + transaction_coins * number_of_periods
+    )
+
+
+def test_exceed_period_limit_should_not_affect_the_next_period(cluster):
+    """
+    check exceeding periodic fee should not affect next period
+    """
+    transaction_coins = 100
+    fee_coins = 10
+
+    period = 5
+    period_limit = 11
+
+    fee_granter_address = cluster.address("community")
+    fee_grantee_address = cluster.address("ecosystem")
+    receiver_address = cluster.address("reserve")
+
+    fee_granter_balance = cluster.balance(fee_granter_address)
+    fee_grantee_balance = cluster.balance(fee_grantee_address)
+    receiver_balance = cluster.balance(receiver_address)
+
+    revoke_fee_grant(cluster, fee_granter_address, fee_grantee_address)
+    grant_fee_allowance(
+        cluster,
+        fee_granter_address,
+        fee_grantee_address,
+        period_limit="%s%s" % (period_limit, BASECRO_DENOM),
+        period=period,
+    )
+
+    transfer(
+        cluster,
+        fee_grantee_address,
+        receiver_address,
+        "%s%s" % (transaction_coins, BASECRO_DENOM),
+        fees="%s%s" % (fee_coins, BASECRO_DENOM),
+        fee_account=fee_granter_address,
+    )
+
+    failed_tx = transfer(
+        cluster,
+        fee_grantee_address,
+        receiver_address,
+        "%s%s" % (transaction_coins, BASECRO_DENOM),
+        fees="%s%s" % (fee_coins, BASECRO_DENOM),
+        fee_account=fee_granter_address,
+    )
+    assert failed_tx["code"] != SUCCESS_CODE, "should fail as fee exceeds period limit"
+    sleep(period)
+
+    transfer(
+        cluster,
+        fee_grantee_address,
+        receiver_address,
+        "%s%s" % (transaction_coins, BASECRO_DENOM),
+        fees="%s%s" % (fee_coins, BASECRO_DENOM),
+        fee_account=fee_granter_address,
+    )
+
+    # transaction only happened two times
+    assert cluster.balance(fee_granter_address) == fee_granter_balance - fee_coins * 2
+    assert (
+        cluster.balance(fee_grantee_address)
+        == fee_grantee_balance - transaction_coins * 2
+    )
+    assert cluster.balance(receiver_address) == receiver_balance + transaction_coins * 2
+
+
+def test_revoke_fee_grant(cluster):
+    """
+    check tx should fail after fee grant is revoked
+    """
+    transaction_coins = 100
+    fee_coins = 10
+
+    fee_granter_address = cluster.address("community")
+    fee_grantee_address = cluster.address("ecosystem")
+    receiver_address = cluster.address("reserve")
+
+    revoke_fee_grant(cluster, fee_granter_address, fee_grantee_address)
+    grant_fee_allowance(cluster, fee_granter_address, fee_grantee_address)
+
+    revoke_fee_grant(cluster, fee_granter_address, fee_grantee_address)
```

Comment:
why is there this double revoke?

Reply:
All test cases execute sequentially but with different feegrant options, the first revoke before the grant removes the previous grant in previous test cases.